Error compiling IPv6ToBinary.cpp

Made sure that `git diff` result is empty

**Operating system**
Mac OS is 10.14.5

**Cmake version**
cmake version 3.14.5

**Ninja version**
1.9.0
=======
/usr/local/bin/g++-9  -DBOOST_SYSTEM_NO_DEPRECATED -DPOCO_STATIC -DPOCO_UNBUNDLED_ZLIB -DUNALIGNED_OK -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -I../libs/libcommon/include -Ilibs/libcommon/include -I../dbms/src -Idbms/src -I../contrib/zlib-ng -Icontrib/zlib-ng -I../contrib/cityhash102/include -I../libs/libwidechar_width -I../contrib/croaring -isystem ../contrib/double-conversion -isystem ../contrib/brotli/c/include -isystem ../contrib/re2 -isystem ../contrib/poco/Data/include -isystem ../contrib/libpcg-random/include -isystem ../contrib/poco/Util/include -isystem ../contrib/poco/Foundation/include -isystem ../contrib/poco/XML/include -isystem ../contrib/poco/JSON/include -isystem ../contrib/boost -isystem ../contrib/poco/Net/include -isystem contrib/re2_st -isystem ../contrib/libcpuid/include -isystem ../contrib/poco/NetSSL_OpenSSL/include -isystem ../contrib/poco/Crypto/include -isystem ../contrib/ssl/include -fdiagnostics-color=always  -msse4.1 -msse4.2 -mpopcnt  -fno-omit-frame-pointer  -Wall  -Wnon-virtual-dtor -Wno-array-bounds -Wextra -Werror -O2 -g -DNDEBUG -O3  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk   -fno-tree-loop-distribute-patterns -std=c++17 -MD -MT dbms/CMakeFiles/clickhouse_common_io.dir/src/Common/IPv6ToBinary.cpp.o -MF dbms/CMakeFiles/clickhouse_common_io.dir/src/Common/IPv6ToBinary.cpp.o.d -o dbms/CMakeFiles/clickhouse_common_io.dir/src/Common/IPv6ToBinary.cpp.o -c ../dbms/src/Common/IPv6ToBinary.cpp
../dbms/src/Common/IPv6ToBinary.cpp: In function 'std::array<char, 16> DB::IPv6ToBinary(const Poco::Net::IPAddress&)':
../dbms/src/Common/IPv6ToBinary.cpp:14:9: error: 'memcpy' was not declared in this scope
   14 |         memcpy(res.data(), address.addr(), 16);
      |         ^~~~~~
../dbms/src/Common/IPv6ToBinary.cpp:3:1: note: 'memcpy' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
    2 | #include <Poco/Net/IPAddress.h>
  +++ |+#include <cstring>
    3 | 
../dbms/src/Common/IPv6ToBinary.cpp:19:9: error: 'memset' was not declared in this scope
   19 |         memset(res.data(), 0, 10);
      |         ^~~~~~
../dbms/src/Common/IPv6ToBinary.cpp:19:9: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
../dbms/src/Common/IPv6ToBinary.cpp:22:9: error: 'memcpy' was not declared in this scope
   22 |         memcpy(&res[12], address.addr(), 4);
      |         ^~~~~~
../dbms/src/Common/IPv6ToBinary.cpp:22:9: note: 'memcpy' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
../dbms/src/Common/IPv6ToBinary.cpp:25:9: error: 'memset' was not declared in this scope
   25 |         memset(res.data(), 0, 16);
      |         ^~~~~~
../dbms/src/Common/IPv6ToBinary.cpp:25:9: note: 'memset' is defined in header '<cstring>'; did you forget to '#include <cstring>'?
[2341/3349] Building CXX object dbms/CMakeFiles/dbms.dir/src/Dictionaries/DictionaryStructure.cpp.o
ninja: build stopped: subcommand failed.

Now I got the following:

FAILED: dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/FunctionsGeo.cpp.o 
/usr/local/bin/g++-9  -DBOOST_SYSTEM_NO_DEPRECATED -DLZ4_DISABLE_DEPRECATE_WARNINGS=1 -DPOCO_STATIC -DPOCO_UNBUNDLED_ZLIB -DUNALIGNED_OK -DWITH_GZFILEOP -DX86_64 -DZLIB_COMPAT -Idbms/src/Functions/include -I../dbms/src -Idbms/src -I../libs/libcommon/include -Ilibs/libcommon/include -I../contrib/zlib-ng -Icontrib/zlib-ng -I../contrib/cityhash102/include -I../contrib/croaring -I../libs/consistent-hashing -I../libs/consistent-hashing-sumbur -I../contrib/libfarmhash -I../contrib/murmurhash/include -I../contrib/simdjson/include -isystem ../contrib/libdivide -isystem ../contrib/libmetrohash/src -isystem /usr/local/opt/icu4c/include -isystem ../contrib/base64/include -isystem ../contrib/lz4/lib -isystem ../contrib/hyperscan/src -isystem ../contrib/rapidjson/include -isystem ../contrib/pdqsort -isystem ../contrib/libpcg-random/include -isystem ../contrib/double-conversion -isystem ../contrib/re2 -isystem ../contrib/poco/Util/include -isystem ../contrib/poco/Foundation/include -isystem ../contrib/poco/XML/include -isystem ../contrib/poco/JSON/include -isystem ../contrib/boost -isystem ../contrib/poco/Net/include -isystem contrib/re2_st -isystem ../contrib/ssl/include -fdiagnostics-color=always  -msse4.1 -msse4.2 -mpopcnt  -fno-omit-frame-pointer  -Wall  -Wnon-virtual-dtor -Wno-array-bounds -Wextra -Werror -O2 -g -DNDEBUG -O3  -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.14.sdk   -fno-tree-loop-distribute-patterns -g0 -std=c++17 -MD -MT dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/FunctionsGeo.cpp.o -MF dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/FunctionsGeo.cpp.o.d -o dbms/src/Functions/CMakeFiles/clickhouse_functions.dir/FunctionsGeo.cpp.o -c ../dbms/src/Functions/FunctionsGeo.cpp
In file included from ../dbms/src/Functions/FunctionsGeo.cpp:3:
../dbms/src/Functions/GeoUtils.h: In member function 'void DB::GeoUtils::PointInPolygonWithGrid<CoordinateType>::buildGrid() [with CoordinateType = float]':
../dbms/src/Functions/GeoUtils.h:228:35: error: '*((void*)& box +12)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  228 |     cell_height = (max_corner.y() - min_corner.y()) / grid_size;
      |                   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
../dbms/src/Functions/GeoUtils.h:245:9: note: '*((void*)& box +12)' was declared here
  245 |     Box box;
      |         ^~~
../dbms/src/Functions/GeoUtils.h:227:34: error: '*((void*)& box +8)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  227 |     cell_width = (max_corner.x() - min_corner.x()) / grid_size;
      |                  ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
../dbms/src/Functions/GeoUtils.h:245:9: note: '*((void*)& box +8)' was declared here
  245 |     Box box;
      |         ^~~
../dbms/src/Functions/GeoUtils.h:228:35: error: '*((void*)& box +4)' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  228 |     cell_height = (max_corner.y() - min_corner.y()) / grid_size;
      |                   ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
../dbms/src/Functions/GeoUtils.h:245:9: note: '*((void*)& box +4)' was declared here
  245 |     Box box;
      |         ^~~
../dbms/src/Functions/GeoUtils.h:227:34: error: 'box' may be used uninitialized in this function [-Werror=maybe-uninitialized]
  227 |     cell_width = (max_corner.x() - min_corner.x()) / grid_size;
      |                  ~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~
../dbms/src/Functions/GeoUtils.h:245:9: note: 'box' was declared here
  245 |     Box box;
      |         ^~~
cc1plus: all warnings being treated as errors

--- a/dbms/src/Functions/GeoUtils.h
+++ b/dbms/src/Functions/GeoUtils.h
@@ -223,9 +223,16 @@ void PointInPolygonWithGrid<CoordinateType>::calcGridAttributes(
 
     const Point & min_corner = box.min_corner();
     const Point & max_corner = box.max_corner();
+   
+#pragma GCC diagnostic push
+#if !__clang__
+#pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
+#endif
 
-    cell_width = (max_corner.x() - min_corner.x()) / grid_size;
-    cell_height = (max_corner.y() - min_corner.y()) / grid_size;
+     cell_width = (max_corner.x() - min_corner.x()) / grid_size;
+     cell_height = (max_corner.y() - min_corner.y()) / grid_size;
+
+#pragma GCC diagnostic pop
 
     if (cell_width == 0 || cell_height == 0)
     {

--- a/dbms/src/Functions/GeoUtils.h
+++ b/dbms/src/Functions/GeoUtils.h
@@ -223,7 +223,7 @@ void PointInPolygonWithGrid<CoordinateType>::calcGridAttributes(
 
     const Point & min_corner = box.min_corner();
     const Point & max_corner = box.max_corner();
-   
+
 #pragma GCC diagnostic push
 #if !__clang__
 #pragma GCC diagnostic ignored "-Wmaybe-uninitialized"

--- a/dbms/src/Functions/GeoUtils.h
+++ b/dbms/src/Functions/GeoUtils.h
@@ -229,8 +229,8 @@ void PointInPolygonWithGrid<CoordinateType>::calcGridAttributes(
 #pragma GCC diagnostic ignored "-Wmaybe-uninitialized"
 #endif
 
-     cell_width = (max_corner.x() - min_corner.x()) / grid_size;
-     cell_height = (max_corner.y() - min_corner.y()) / grid_size;
+    cell_width = (max_corner.x() - min_corner.x()) / grid_size;
+    cell_height = (max_corner.y() - min_corner.y()) / grid_size;
 
 #pragma GCC diagnostic pop
 

